plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializerIrGenerator.kt

```diff
@@ -491,17 +521,103 @@ open class SerializerIrGenerator(
-            serializableSyntheticConstructor(serializableIrClass)
+            val ctor: IrConstructorSymbol = serializableSyntheticConstructor(serializableIrClass)
+            +irReturn(irInvoke(null, ctor, typeArgs, args))
         } else {
-            compilerContext.referenceConstructors(serializableDescriptor.fqNameSafe).single { it.owner.isPrimary }
+            // check properties presence
+            if (useFieldMissingOptimization) {
+                val irProperty = irClass.searchForProperty(generatedSerialDescPropertyDescriptor!!)
+                val serialDescriptorExpr = irGetField(null, irProperty.backingField!!)
+                generateGoldenMaskCheck(bitMasks, properties, serialDescriptorExpr)
+            }
+
+            val ctor: IrConstructorSymbol =
+                compilerContext.referenceConstructors(serializableDescriptor.fqNameSafe).single { it.owner.isPrimary }
+            val params = ctor.owner.valueParameters
+
+
+            val paramReplacer: (ValueParameterDescriptor) -> IrExpression? = {
+                val propertyDescriptor = bindingContext[BindingContext.VALUE_PARAMETER_AS_PROPERTY, it]
+                if (propertyDescriptor != null) {
+                    val serializable = serialPropertiesMap[propertyDescriptor]
+                    (serializable ?: transientsPropertiesMap[propertyDescriptor])?.get()
+                } else {
+                    null
+                }
+            }
+            val prepareInitializer: (IrExpressionBody) -> IrExpression = createInitializerPreparer(serializableIrClass, paramReplacer)
+
+            // constructor args:
+            val ctorArgs = params.map { parameter ->
+                val parameterDescriptor = parameter.descriptor as ValueParameterDescriptor
+                val propertyDescriptor = bindingContext[BindingContext.VALUE_PARAMETER_AS_PROPERTY, parameterDescriptor]!!
+                val serialProperty = serialPropertiesMap[propertyDescriptor]
+
+                // null if transient
+                if (serialProperty != null) {
+                    val index = serialPropertiesIndexes.getValue(propertyDescriptor)
+                    if (parameterDescriptor.hasDefaultValue()) {
+                        val propNotSeenTest =
+                            irEquals(
+                                irInt(0),
+                                irBinOp(
+                                    OperatorNameConventions.AND,
+                                    bitMasks[index / 32].get(),
+                                    irInt(1 shl (index % 32))
+                                )
+                            )
+
+                        // if(mask$j && propertyMask == 0) local$i = <initializer>
+                        val defaultValueExp = parameter.defaultValue!!
+                        val expr = prepareInitializer(defaultValueExp)
+                        +irIfThen(propNotSeenTest, irSet(serialProperty.symbol, expr))
```

Comment:
I do not see any `throw` calls in other branch. Do you assume that they all were generated in `generateGoldenMaskCheck` ? Then probably we do not need an `if (useFieldMissingOptimization)` — as I said, this check may fail because IR klibs do not have implementation-version in manifest

Reply:
In this case, external serializers can be compiled only with library >= 1.1.0-M1. Is it ok?

Reply:
I think it's ok — external serializers are an experimental feature anyway